### ISSUE TITLE
Support more schema collections in GetSchema

### ISSUE THREAD

**Issue by bgrainger** [mysql-net/MySqlConnector#375]
Connector/NET supports the following schema collections in `MySqlConnection.GetSchema`. MySqlConnector only supports the bolded entries.

* **MetaDataCollections**
* DataSourceInformation
* **DataTypes**
* Restrictions
* ReservedWords
* Databases
* Tables
* Columns
* Users
* Foreign Keys
* IndexColumns
* Indexes
* Foreign Key Columns
* UDF
* Views
* ViewColumns
* Procedure Parameters
* **Procedures**
* Triggers

Add support for additional schema collections that weren't implemented in #361.

**Comment by phillipc** [comment]
Look at my PR: I have implemented Foreign Keys-, IndexColumns-, Indexes- and Foreign Key Columns-Support

**Comment by eaexo** [comment]
Hi @bgrainger

Are there any plans to fix this issue? Missing schema collections make the mysqlconnector.net kind of useless with NHibernate (even if one could discuss if it make sense to use automatic scheme update)

**Comment by bgrainger** [comment]
`Foreign Keys`, `Indexes`, and `Index Columns` support were added by @eaexo in #1419.

**Comment by eaexo** [comment]
Hello @bgrainger 
when can we expect to have at least a beta released resp. published to the nuget.org?

**Comment by bgrainger** [comment]
Foreign Keys, Indexes, Index Columns added in [2.3.2](https://github.com/mysql-net/MySqlConnector/releases/tag/2.3.2).

**Comment by eaexo** [comment]
Hi @bgrainger 

unfortunatelly in https://github.com/mysql-net/MySqlConnector/commit/8748a725e6de168c01094f10b6206a8d6a6fae70#diff-b53ea1b55d852cbcb0fe21dab331dcf093424734bd32164bc03f4df0d6624b3cL843 (probably) you broke it again.

At least NHibernate passes 5 restrictions for determining index columns, see  `NHibernate.Dialect.Schema.AbstractDataBaseSchema`.

```

		public virtual DataTable GetIndexColumns(string catalog, string schemaPattern, string tableName, string indexName)
		{
			if (UseDialectQualifyInsteadOfTableName)
			{
				var actualTableName = GetActualTableName(catalog, schemaPattern, tableName);
				return Connection.GetSchema("IndexColumns", new[] {null, null, actualTableName, indexName, null});
			}

			var restrictions = new[] {catalog, schemaPattern, tableName, indexName, null};
			return Connection.GetSchema("IndexColumns", restrictions);
		}

```

In my implenentation I had a "placeholder" column which matched the 5th restriction:
`default: COLUMN_NAME`.

Should I re-add it again, so it works, or should a different mechanism be introduced to allow more restrictions then really needed, e.g. by introducing something like boolean placeholder flag, which says, that the restriction is there but is not mapped to a specific column?

**Comment by bgrainger** [comment]
Yes, I removed the placeholder that you added with no implementation and no tests.

Added it back with an implementation and working tests: https://github.com/mysql-net/MySqlConnector/commit/d3ad6d1acf8f70e0e693ac845e3ed342222396c1.

**Comment by eaexo** [comment]
@bgrainger 

Thanks for quick update.

**Comment by eaexo** [comment]
@bgrainger 
If you like I could try to implement a better solution for "placeholders" in restrictions, so we do not need to specify a "random" column, but saying that any argument starting with index>n will be ignored, or that one can specify a minimum amount of restriction items

**Comment by bgrainger** [comment]
The better solution is implementing the restriction, which I've done: https://github.com/mysql-net/MySqlConnector/commit/d3ad6d1acf8f70e0e693ac845e3ed342222396c1#diff-f0b4261dcdfed8eaab87870a6bce4b45f0a6c91aadddb44b47ff239665b4a082R535-R539

**Comment by nikkilocke** [comment]
Would you like some contributions to this? I could do with Foreign Key Columns and ViewColumns, for example. Also, it would be good to add some documentation on how restrictions work (as these are very different to the MySql.Net ones). I figured it out by looking at the source, but that shouldn't really be necessary, should it?
If I find time to do some of it, do you have guidelines on code style, tests, etc.?

**Comment by bgrainger** [comment]
> Also, it would be good to add some documentation on how restrictions work (as these are very different to the MySql.Net ones). I figured it out by looking at the source, but that shouldn't really be necessary, should it?

Is the current documentation site insufficient? https://mysqlconnector.net/overview/schema-collections/ and e.g., https://mysqlconnector.net/overview/schema/tables/ (It assumes you're familiar with the background info from MSDN: https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/schema-restrictions).

> Would you like some contributions to this? I could do with Foreign Key Columns and ViewColumns, for example.

Yes, PRs are welcome.

> If I find time to do some of it, do you have guidelines on code style, tests, etc.?

`.editorconfig` in the repo should already set code style (in any editor that respects that). Please add tests to https://github.com/mysql-net/MySqlConnector/blob/master/tests/IntegrationTests/SchemaProviderTests.cs, following the general pattern in that file.

**Comment by nikkilocke** [comment]
Thanks - I hadn't found that documentation - I usually expect documentation on methods to be in `///` comments above the method, so Visual Studio displays it without having to go to a website. Will read it, now I know where it is.

Will do a PR if I get time (breath holding is not advised, as I'm a bit busy!).

**Comment by BogdanYarotsky** [comment]
Hi, I will look into GetSchema() implementation for 'Users' and will try to compose a PR in the following weeks.
